### DISCUSSION TITLE
CSV rendering bug in GitHub Web App

### DISCUSSION THREAD

**Discussion by SteveALee** [community/community#22405] [New to GitHub]
<p>[I’d raise a bug but this seems to be the option I have, hope is correct tag for the Web App]</p>
<p>I have a CSV file that uses <code>;</code> as is common in EU countries that use <code>,</code> as the currency separator. As far as I understand the CSV ‘specification’ that allows <code>,</code> to be used in cells.</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/mulberrysymbols/mulberry-symbols/blob/7279a34774d11af63ec14489573f538041c072d7/scripts/data/FinnishSymbolNames.csv">
  <header class="source">

      <a href="https://github.com/mulberrysymbols/mulberry-symbols/blob/7279a34774d11af63ec14489573f538041c072d7/scripts/data/FinnishSymbolNames.csv" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/mulberrysymbols/mulberry-symbols/blob/7279a34774d11af63ec14489573f538041c072d7/scripts/data/FinnishSymbolNames.csv" target="_blank" rel="noopener nofollow ugc">mulberrysymbols/mulberry-symbols/blob/7279a34774d11af63ec14489573f538041c072d7/scripts/data/FinnishSymbolNames.csv</a></h4>

    <pre><code class="lang-csv">﻿symbol-id;symbol-en;symbol-fi;category
1;A;A;Aakkoset
2;a_-_lower_case;a;Aakkoset
3;above;yläpuolella;Kuvaileva sijainti
4;absent;poissa;Kuvaileva tila
5;acne_spots;akne;Terveys terveydentila
6;acorn;tammenterho;Kasvit ja puut
7;across;yli;Kuvaileva suunta
8;activity_centre;toimintakeskus;Vapaa-aika lelut
9;add;plus;Numerot laskutoimitukset
10;Advent;adventti;Uskonto juhlapäivä
11;Advent_calendar;adventtikalenteri;Uskonto yleiset
12;aerial_indoor;sisäantenni;Tekniikka TV
13;aerobics_dance;aerobic;Urheilu
14;aeroplane;lentokone;Liikenne ilmailu
15;afraid_lady;nainen peloissaan;Ihmiset tunteet
16;afraid_man;mies peloissaan;Ihmiset tunteet
17;after;perässä;Kuvaileva sijainti
18;afternoon;iltapäivä;Kuvaileva aika
19;aftershave;partavesi;Terveys hoitotarvikkeet
</code></pre>

  This file has been truncated. <a href="https://github.com/mulberrysymbols/mulberry-symbols/blob/7279a34774d11af63ec14489573f538041c072d7/scripts/data/FinnishSymbolNames.csv" target="_blank" rel="noopener nofollow ugc">show original</a>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>However that seems to be breaking the GitHub pretty rendering of the CSV.</p>
<pre><code class="lang-auto">We can make this file beautiful and searchable if this error is corrected: It looks like row 56 should actually have 1 column, instead of 2. in line 55. 
</code></pre>
<p>For these lines</p>
<pre><code class="lang-auto">54;arms;kädet;Terveys kehonosat
55;armwrestle_,_to;kädenvääntö;Vapaa-aika pelit
</code></pre>
<p>Is this a CSV parsing bug.? A result of not having a real CSV standard? A localisation assumption bug? Or something else?</p>
<p>Can I work around it?<br>
A hack is probably to convert the CSV to <code>,</code> separated and quoting strings with <code>,</code>s but would rather keep as supplied to me by a contributor.</p>
<p>Thanks</p>
<p>Steve</p>

**Comment by HemantSachdeva** [comment]
<aside class="quote no-group" data-username="SteveALee" data-post="1" data-topic="212487">
<div class="title">
<div class="quote-controls"></div>
 SteveALee:</div>
<blockquote>
<p>A hack is probably to convert the CSV to <code>,</code> separated and quoting strings with <code>,</code> s but would rather keep as supplied to me by a contributor.</p>
</blockquote>
</aside>
<p>This is the only solution.</p>
<p>The reason why only that line is showing error is because it have <code>,</code><br>
If none of your data had <code>,</code> then you will see the correct error message:</p>
<p><code>We can make this file beautiful and searchable if this error is corrected: No commas found in this CSV file in line 0.</code></p>

**Comment by SteveALee** [comment]
<p>Shame as <code>;</code> is a valid separator in many places CSV get used.</p>
<p>But again, there’s no real <a href="https://en.wikipedia.org/wiki/Comma-separated_values#Standardization" rel="noopener nofollow ugc">CSV standard</a> so we can’t complain. Sign.</p>

**Comment by HemantSachdeva** [comment]
<p>I don’t see anything wrong in only accepting <code>,</code> as delimiter as name says <code>Comma Separated Values</code>.</p>

**Reply by LWChris** [reply]
A classical "Sounds like a You-Problem to me." answer 😂

A substantial part of the world uses the comma (light green) as decimal separator instead of the period (blue).

![Decimal separator world map](https://upload.wikimedia.org/wikipedia/commons/a/a8/DecimalSeparator.svg)
Source: NuclearVacuum, [CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0), via Wikimedia Commons

When you save a CSV, say, from Excel, it uses semicolons as delimiters. Even worse, when you open a CSV file that actually uses commas, Excel will not split the content into columns. You need to use the "Split into Columns" tool to get a proper table out - which has its own Excel-typical downsides like Excel messing with the source data while trying to format the columns etc.

You can't tell Excel to use the other one, because it uses the Windows locale, so you'd have to set Windows to English to get this fixed.

That's why it's impractical to use "," as delimiter in about half of the World. And since whoever at Microsoft thought it might be a good idea to localize the content but didn't think as far as assigning "ssv" as extension, all those people are stuck with a file name extension that says "comma" and contents that use semicolon, unless you manually format it back and forth every time after you save/before you open.

I would like to second this request, and since the feature is not there, neither is the feature request linked in the post marked answer (clicking the link opens the overview of all requests, so apparently that one was deleted), I'm bumping this thread again.

**Comment by koppor** [comment]
Note: W3C also lists available "standads" at https://w3c.github.io/csvw/syntax/#standards. Unfortunately, all of them say `,` is the stanard.

**Reply by LWChris** [reply]
> Note: W3C also lists available "standads" at https://w3c.github.io/csvw/syntax/#standards. Unfortunately, all of them say `,` is the stanard.

Yes, but it acknowledges that this is a "non-normative" "attempt" made with "Excel for Mac 2011" set to an "English locale". So it should be treated about as binding as an HTML standard derived from analysis of HTML files that Internet Explorer 8 can render correctly.

It also says that the line ending should be CRLF. Following that logic GitHub should also refuse to render CSV with LF line endings.

To say it with the words of [section 3 in RFC 4180](https://www.rfc-editor.org/rfc/rfc4180#section-3), which lends itself from section 2.10. Robustness Principle of the RFC 793 (TCP):
> be conservative in what you do, be liberal in what you accept from others

Sadly, it appears that GitHub chooses to ignore that sentiment with a "works for me" attitude.

**Comment by Siedlerchr** [comment]
Back in 2019 there was still the discourse forum. Unfortunately, GitHub did not migrate the stuff correctly to the discussions, so the link above in the marked as resolved is dead. The thread had 100s of upvotes.

Although the standard says it's comma, a semicolon is stil very common in Europe and default when exporting from e.g. German excel.
S

**Comment by koppor** [comment]
I found a very good instruction to change Excel's behavior at https://answers.microsoft.com/en-us/msoffice/forum/all/how-to-change-semicolon-to-comma-when-saving-csv/cf7c9537-82da-448d-9516-82f3a4b13fae.

I know that it is not feasible to request that change from nearly the half of all Excel users of the world.

1. Click on the Start menu.
2. Type control panel.
3. Click on Control Panel (Desktop app).
4. Select Clock, Language, and Region.
5. Click Region.
6. Select Additional settings in the pop-up window.
7. Change the List separator to comma.
8. Click Apply.
9. Hit OK.

**Comment by LWChris** [comment]
Standards and would-be standards aside, I really don't understand what is so difficult about just doing it. Feels like people are rather opting into years of debate whether or not to be considerate about others, than into investing a few days of coding to be.

Whole VS code extensions with that functionality (and more) have been created whilst this discussion was ongoing.

So who is responsible, who could add that feature? A decent heuristic algorithm shouldn't be hard to develop.

**Reply by LWChris** [reply]
Thinking about it, if the comma heuristic fails, take the first two lines of the CSV file, and count the number of semicolons for both lines. If the count matches and is greater than 0, assume semicolon as separator.

That should be a very defensive approach. It cannot make anything worse for existing cases since it only kicks in when the current heuristic has given up. But it should work for many cases.

**Reply by ChristianKollross** [reply]
Or allow users to set a default separator per project. (And then maybe apply the heuristic.)

**Reply by LWChris** [reply]
Yes, but allowing to set a default would be an entirely new feature. It would be sensible, but given that even the least-invasive approach I mentioned is ignored, I wouldn't assume a more involved approach would see any more traction, sadly.

**Comment by HotCakeX** [comment]
In all honesty, 
CSV = `Comma Separated Values`, **not** `Semi-colon separated Values`

it's an easy find & replace to correct CSVs with delimiters other than comma.

**Reply by LWChris** [reply]
It appears that you have not read the thread. TLDR for you: in files that are using semicolon, the comma is used for decimals. Changing the format breaks compatibility with Excel and other Microsoft software.

The file name extension "*.csv" has been assigned to both the English format using commas and the other format using semicolons. There is nothing we can do about that, it is a misnomer, but it's not a mistake a handful of people have made, but the official behavior of the most popular software used to create these files and therefore a de-facto standard.

Sure, if the world was a different one where Microsoft had decided to use "*.ssv", I would agree. But the reality is that for 50% of the world, clicking "File > Save as > CSV" in Excel will create a file using semicolons, and opening a *.csv that uses commas will not work.

**Reply by HotCakeX** [reply]
Thank you for the explanation, but I'm curious, why convert an Excel file to CSV in the first place? Excel can stay in .xlsx file format and be opened anywhere, phones, computers, tablets, laptops, literally anywhere with a browser. It's a rich file format compared to CSV which is text based.

**Reply by SteveALee** [reply]
To import data into other programs. As a "standard: er cough, CSV allows interchange between all programs that follow the standard. 

CSV is a so called "defacto standard", it's evolved from usage rather than having a full officially maintained specification. Hence this thread. :)

**Reply by LWChris** [reply]
Or to put it in a different way: CSV is meant to be a simple, low overhead text-only format for tabular (table-shaped) data. And since Excel is Microsoft's editor for tabular data, Excel is the default program to open, view, edit and create CSV files for most people.

Due to its wide spread usage Excel has defined a de-facto standard of what a CSV file should look like.

And there are two unfortunate things here:

a) Most computer stuff including standards being conceived in America being all too quick in conveniently forgetting about the rest of the world, so a lot of standardization working perfectly fine for American culture but not so much for other cultures.

b) Microsoft apparently decided that human readability in a text editor is important. So rather than defining "." as the fixed character for the decimal point and "," as data separator, they went with localized number format (which uses "," as decimal point for 50% of the world) and therefore needed to also go with a localized data separator: ";".

Microsoft's interpretation and implementation of the CSV format is a notable exception to a), whereas apparently all "modern" retrofitted standardization attempts of CSV fail to acknowledge that again.

**Reply by SteveALee** [reply]
Much truth. Just look back at ASCII, and then MS having to introduce the code pages hack to MSDOS for localisation/internationalisation (s spelling  intentional - snigger).

Of course the tell tale is in the full name that ASCII is short for! Lol

**Comment by DavidClaszen** [comment]
So is there a way to un-answer this and get this back on the radar again somehow? I don't often look at my CSV files up on Github, but that's at least partly due to the fact that they're all garbled. Are Europeans simply not allowed to have pretty formatted data on Github?

**Reply by SteveALee** [reply]
I wonder how many people in the same boat!

**Reply by Siedlerchr** [reply]
I deleted my comment so it should be marked as unanswered now

**Reply by LWChris** [reply]
So, I felt this fish tank is going nowhere. I feel like we're all just sitting in here, talking to ourselves about how ridiculous this situation is, but nobody from the outside (i.e. GitHub product team) notices. So I contacted GitHub support and offered them a free patch to their product code if they could connect me with whoever can work on this stuff.

Guess what, their reply was basically "Thank you for contacting us, but this is the support team, this stuff should be discussed with the product team. Best to add a discussion in the community discussions."

So I guess there's only two things we can do:

1) properly label this thread with "labels: Product Feedback, Enhancement, Accessibility, Bug" or whatever fits
2) tell everyone we know who has the same problem about this topic, and get them all to upvote it, so it will hopefully finally get the attention by someone.

Until then, we can keep making bubbles in our fish tank in here.

**Comment by tom9358** [comment]
I have some Dutch csv files that are semicolon-separated and it would be fantastic if the detection of csv attempted to separate by semicolon instead of stoppin with the message "We can make this file [beautiful and searchable](https://docs.github.com/articles/rendering-csv-and-tsv-data) if this error is corrected"! Really hope someone picks this up. VSCode supports it just fine, but I work with people who would like to manually inspect the files in the web view
